deploy-service/teletraanservice/src/main/java/com/pinterest/teletraan/worker/AutoPromoter.java

```diff
@@ -122,32 +123,116 @@ boolean isCurrentDeployRetirable(String envId, DeployBean currDeployBean) throws
+            StringUtils.isNotEmpty(result.getPromotedBuild())){
+            safePromote(null,  result.getPromotedBuild(), Constants.BUILD_STAGE, currDeployBean,
+                currEnvBean);
         }
     }
 
-    void promoteDeploy(EnvironBean currEnvBean, DeployBean currDeployBean, int size, PromoteBean predPromoteBean) throws Exception {
+    //This contains the logic about if there should be a promote deploy from the pred environment.
+    public PromoteResult computePromoteDeployResult(EnvironBean currEnvBean, DeployBean currDeployBean,
+                                                    int size, PromoteBean predPromoteBean) throws Exception {
         String predStage = predPromoteBean.getPred_stage();
         // Special case when there is no pred deploy
         EnvironBean predEnvBean = environDAO.getByStage(currEnvBean.getEnv_name(), predStage);
         if (predEnvBean == null) {
-            LOG.warn("Pred env {}/{} does not exist, bail out!", currEnvBean.getEnv_name(), predStage);
-            return;
+            LOG.warn("Pred env {}/{} does not exist, bail out!", currEnvBean.getEnv_name(),
+                predStage);
+            return new PromoteResult()
+                .withResultCode(PromoteResult.ResultCode.NoPredEnvironment);
         }
 
         String predDeployId = predEnvBean.getDeploy_id();
         if (predDeployId == null) {
-            LOG.debug("Pred env {}/{} does not have deploy yet, bail out!", currEnvBean.getEnv_name(), predStage);
-            return;
+            LOG.debug("Pred env {}/{} does not have deploy yet, bail out!",
+                currEnvBean.getEnv_name(), predStage);
+            return new PromoteResult().withResultCode(PromoteResult.ResultCode.NoPredEnvironmentDeploy);
+        }
+
+        Date now = new Date();
+        long mustBefore = Long.MAX_VALUE;
+        String schedule = predPromoteBean.getSchedule();
+        if (!StringUtils.isEmpty(schedule)) {
+            long start_date = 0;
+            if (currDeployBean!=null){
+                start_date = currDeployBean.getStart_date();
+            }
+            else{
+                //If no deploy in current environment, use the time for the pred
+                start_date = deployDAO.getById(predDeployId).getStart_date();
+            }
+
+            //Get the time that we should deploy per schedule
+            Date autoDeployDueDate = getScheduledCheckDueTime(start_date, schedule);
+            if (autoDeployDueDate.getTime() != 0) {
+                mustBefore = autoDeployDueDate.getTime();
+            }
+            if (now.before(autoDeployDueDate)) {
+                return new PromoteResult()
+                    .withResultCode(PromoteResult.ResultCode.NotInScheduledTime);
+            }
+            // For cron auto promote, we only choose the latest one.
+            size = Math.min(size, 1);
         }
 
         long currentDeployDate = 0;
```

Comment:
I do not see currentDeployDate  ever changes its value.  If you use interval, then you do not need this zero-valued variable.